service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AgentDirectoryExtractor.java

```diff
@@ -0,0 +1,232 @@
+        return bundle.getEntry()
+            .stream()
+            .map(Bundle.BundleEntryComponent::getResource)
+            .filter(resource -> resource.getResourceType().equals(ResourceType.Observation))
+            .map(Observation.class::cast)
+            .filter(Observation::hasPerformer)
+            .filter(AgentDirectoryExtractor::performerContainersOrgAndPractitioner)
+            .collect(Collectors.toList());
+    }
+
+    public static Optional<Patient> extractPatientByNhsNumber(Bundle bundle, String nhsNumber) {
+        return bundle.getEntry()
+            .stream()
+            .map(Bundle.BundleEntryComponent::getResource)
+            .filter(AgentDirectoryExtractor::isPatientResource)
+            .map(Patient.class::cast)
+            .filter(patient -> isNhsNumberMatching(patient, nhsNumber))
+            .findFirst();
+    }
+
+    public static List<Practitioner> extractRemainingPractitioners(Bundle bundle) {
+        return bundle.getEntry()
+            .stream()
+            .map(Bundle.BundleEntryComponent::getResource)
+            .filter(resource -> REMAINING_RESOURCE_TYPES.contains(resource.getResourceType()))
+            .filter(AgentDirectoryExtractor::containsRelevantPractitioner)
+            .map(AgentDirectoryExtractor::extractIIdTypes)
+            .map(reference -> ResourceExtractor.extractResourceByReference(bundle, reference))
+            .flatMap(Optional::stream)
+            .map(Practitioner.class::cast)
+            .distinct()
+            .collect(Collectors.toList());
+    }
+
+    public static List<Triple<Practitioner, Optional<PractitionerRole>, Optional<Organization>>>
+        extractPractitionerRoleTriples(Bundle bundle, List<Practitioner> practitioners) {
+
+        List<PractitionerRole> practitionerRoles = bundle.getEntry()
+            .stream()
+            .map(Bundle.BundleEntryComponent::getResource)
+            .filter(resource -> resource.getResourceType().equals(ResourceType.PractitionerRole))
+            .map(PractitionerRole.class::cast)
+            .filter(practitionerRole -> referencesPractitioner(practitionerRole, practitioners))
+            .collect(Collectors.toList());
+
+        List<Organization> organizations = practitionerRoles.stream()
+            .map(practitionerRole -> practitionerRole.getOrganization().getReferenceElement())
+            .map(reference -> ResourceExtractor.extractResourceByReference(bundle, reference))
+            .filter(Optional::isPresent)
+            .map(Optional::get)
+            .map(Organization.class::cast)
+            .collect(Collectors.toList());
+
+        return practitioners.stream()
+            .map(practitioner -> combineTriple(practitioner, practitionerRoles, organizations))
+            .collect(Collectors.toList());
+    }
+
+    private static Triple<Practitioner, Optional<PractitionerRole>, Optional<Organization>> combineTriple(Practitioner practitioner,
```

Comment:
I'm not too keen on the use of `Triple` here, as its members lose semantic value. The `Optional` fields are also questionable. We could create a simple class to encapsulate these data like:
```
    @Data
    public static class AgentData {
        private final Practitioner practitioner;
        private final PractitionerRole practitionerRole;
        private final Organization organization;
    }
```
Then this line can be
```suggestion
    private static AgentData createAgentData(Practitioner practitioner,
```
and the client can refer to `practitioner`, `practitionerRole` and `organization` instead of `left`, `middle` and `right.

You could add getters to the class to return `Optional` still, but otherwise it's just changing things like `if (optional.isPresent()) {` to `if (nullable != null) {`